odoo/addons/base/tests/test_expression.py

```diff
@@ -767,16 +768,211 @@ def test_30_normalize_domain(self):
         with self.assertRaises(ValueError):
             normalize_domain([('a', '=', 1), '!'])
 
+    def test_30_instantiate_domain(self):
+        simple = Domain('foo', '=', 'bar')
+        self.assertIsInstance(simple, Domain, "Domain constructor must be instance of Domain")
+        simple_list = [('foo', '=', 'bar')]
+        simple_domain = Domain(simple_list)
+        self.assertEqual(simple_domain, simple)
+        self.assertIs(Domain(simple), simple, "Domain(Domain) should return the instance")
+        self.assertEqual(Domain(simple_list[0]), simple, "Constructor should work for a tuple")
+
+        # negative and nary
+        neg_domain = ~Domain('foo', '=like', 'bar')
+        assert next(iter(neg_domain)) == "!", "Internal test that we are inversing a domain"
+        and_domain = simple & Domain('bar', '=', 'baz')
+
+        # bool
+        self.assertEqual(Domain(True), Domain.TRUE)
+        self.assertEqual(Domain([]), Domain.TRUE)
+        self.assertEqual(Domain(False), Domain.FALSE)
+
+        # truth value
+        for dom, is_bool in [
+            (simple, None),
+            (Domain.TRUE, True),
+            (Domain.FALSE, False),
+            (and_domain, None),
+            (neg_domain, None),
+        ]:
+            self.assertEqual(bool(dom), dom != Domain.TRUE, "Only TRUE is falsy because it's equivalent to empty domain")
```

Comment:
So this means that
```py
assert not Domain.TRUE
```
Doesn't it look confusing to you?  That's insane, isn't it?
I think that our only sane option is to make `bool(some_domain)` an error.  What do you think?

Reply:
This is to be backward-compatible.

```python
if domain:
   something
```

Something is done when domain is not `[]`. If I replace `domain` by `Domain(domain)` I want the same behaviour as earlier. If you view the domain as a restriction on recordsets, this actually makes sense: an empty restriction (`TRUE`) is falsy. Another sane option would be not to overwrite bool at all.

In any case, there is a lot of patterns like `domain or xxx` where you want `xxx` and you don't know whether domain is a list or a domain object. If we want to break this, I would do it in another PR.

Reply:
I do very much. You should be able to use this class anywhere you had a domain as a list.
I think it is important to have code that now may get a Domain class and still iterates, appends or does other things with it and still continues to work - you just cannot mutate it anymore.